File: packages/template-python/src/deepagents_template/runtime/agent_config.py
"""Agent config parts builder — composes the configuration passed to the agent.

Port of the TS ``src/runtime/agent-config.ts``. Instead of building a
``createDeepAgent`` payload (LangGraph), this assembles the keyword
arguments for a pydantic-ai ``Agent`` constructor: model, system prompt,
tools, model settings, skills, memory, subagents, permissions, and
middleware-compatible hooks.
"""

from __future__ import annotations

from typing import Any

from pydantic_ai import ModelSettings

from deepagents_template.runtime.config.config_schema import ACPSessionConfig, AppConfig
from deepagents_template.runtime.helpers import (
    discover_memory_files as _helpers_discover_memory,
)
from deepagents_template.runtime.helpers import (
    discover_sub_agents as _helpers_discover_sub_agents,
)
from deepagents_template.runtime.helpers import (
    resolve_skills_paths,
    with_runtime_context_prompt,
)
from deepagents_template.runtime.logger import logger
from deepagents_template.runtime.middleware.compaction import create_compaction_middleware
from deepagents_template.runtime.middleware.cost_tracking import create_cost_tracking_middleware
from deepagents_template.runtime.middleware.eviction import create_eviction_middleware
from deepagents_template.runtime.middleware.fs_path_resolver import create_fs_path_resolver
from deepagents_template.runtime.middleware.harness_lifecycle import (
    create_harness_lifecycle_middleware,
)
from deepagents_template.runtime.middleware.periodic_reminder import (
    create_periodic_reminder_middleware,
)
from deepagents_template.runtime.middleware.protected_paths import (
    create_protected_paths_middleware,
)
from deepagents_template.runtime.middleware.stuck_loop import create_stuck_loop_middleware
from deepagents_template.runtime.model import resolve_model, resolve_summarizer_model
from deepagents_template.runtime.permissions import (
    build_interrupt_on,
    build_permissions,
    resolve_sandbox_policy,
    to_absolute_deny_glob,
)
from deepagents_template.runtime.prompt import resolve_system_prompt


def build_agent_config_parts(
    config: AppConfig,
    workspace_root: str,
    tools: list[dict[str, Any]],
    *,
    session_config: ACPSessionConfig | None = None,
    checkpointer: Any = None,
) -> dict[str, Any]:
    """Compose pydantic-ai Agent keyword arguments from ``AppConfig``.

    Returns a dict with keys that can be unpacked into ``Agent(**parts)``:

    * ``model`` — resolved pydantic-ai Model
    * ``system_prompt`` — final system prompt (with runtime context)
    * ``tools`` — list of pydantic-ai ``Tool`` objects
    * ``model_settings`` — ``ModelSettings`` (temperature, max_tokens, …)
    * ``result_type`` — optional structured result schema
    * ``before_model`` / ``after_model`` — hook lists for middleware
    * ``before_tool`` / ``after_tool`` — tool-lifecycle hooks
    """
    log = logger.child("agent-config")
    mw_config = config.middleware

    # ── Model ──────────────────────────────────────────────────────────
    model = resolve_model(config)

    # ── Model Settings ─────────────────────────────────────────────────
    model_settings_kw: dict[str, Any] = {}
    if config.model.settings.temperature is not None:
        model_settings_kw["temperature"] = config.model.settings.temperature
    if config.model.settings.max_tokens is not None:
        model_settings_kw["max_tokens"] = config.model.settings.max_tokens
    model_settings = ModelSettings(**model_settings_kw) if model_settings_kw else None

    # ── System Prompt ──────────────────────────────────────────────────
    system_prompt = with_runtime_context_prompt(
        resolve_system_prompt(config, session_config, workspace_root=workspace_root),
        workspace_root=workspace_root,
    )

    # ── Mode-based overrides ──────────────────────────────────────────
    mode = config.permissions.mode or "ask"
    interrupt_on = build_interrupt_on(config.permissions.interrupt_on or [])
    permissions = build_permissions(config, workspace_root)

    if mode == "plan":
        plan_preamble = (
            "## Planning Mode\n"
            "Before making any changes, you MUST:\n"
            "1. Present a clear plan of what you intend to do\n"
            "2. Wait for user approval\n"
            "3. Only then proceed with execution\n\n"
        )
        system_prompt = plan_preamble + system_prompt
    elif mode == "yolo":
        interrupt_on = {}

    # ── Middleware chain ───────────────────────────────────────────────
    middleware_hooks: dict[str, list[Any]] = {
        "before_model": [],
        "after_model": [],
        "before_tool": [],
        "after_tool": [],
    }

    # Harness lifecycle — always first
    middleware_hooks["before_model"].append(create_harness_lifecycle_middleware())

    # Stuck loop detection
    if mw_config.stuck_loop_detection.enabled:
        middleware_hooks["after_tool"].append(
            create_stuck_loop_middleware(
                threshold=mw_config.stuck_loop_detection.threshold,
                mode=mw_config.stuck_loop_detection.mode,
            )
        )

    # FS path resolver
    middleware_hooks["before_tool"].append(create_fs_path_resolver(workspace_root))

    # Periodic reminder
    if mw_config.periodic_reminder.enabled:
        middleware_hooks["before_model"].append(
            create_periodic_reminder_middleware(
                first_at=mw_config.periodic_reminder.first_at,
                every=mw_config.periodic_reminder.every,
            )
        )

    # Cost tracking
    if mw_config.cost_tracking.enabled:
        middleware_hooks["after_model"].append(
            create_cost_tracking_middleware(
                warn_at_tokens=mw_config.cost_tracking.warn_at_tokens,
            )
        )

    # Compaction (summarization)
    if config.compaction.enabled:
        middleware_hooks["before_model"].append(
            create_compaction_middleware(
                config=config.compaction,
                model=resolve_summarizer_model(config),
            )
        )

    # Eviction
    if config.eviction.enabled:
        middleware_hooks["after_tool"].append(
            create_eviction_middleware(config=config.eviction)
        )

    # Protected paths
    sandbox = resolve_sandbox_policy(config, workspace_root)
    if sandbox.denied_write_paths:
        denied_globs = [
            to_absolute_deny_glob(d, workspace_root)
            for d in sandbox.denied_write_paths
        ]
        middleware_hooks["before_tool"].append(
            create_protected_paths_middleware(denied_globs=denied_globs)
        )

    # ── Skills & Memory ───────────────────────────────────────────────
    skills_paths = resolve_skills_paths(config)
    memory_paths = _helpers_discover_memory(
        workspace_root, config.agent.include_workspace_instructions
    )
    sub_agents = _helpers_discover_sub_agents(config, workspace_root)

    # ── Assemble parts ────────────────────────────────────────────────
    parts: dict[str, Any] = {
        "model": model,
        "system_prompt": system_prompt,
        "tools": tools,
        "model_settings": model_settings,
        "middleware_hooks": middleware_hooks,
        "skills_paths": skills_paths,
        "memory_paths": memory_paths,
        "sub_agents": sub_agents,
        "permissions": permissions,
        "interrupt_on": interrupt_on,
        "checkpointer": checkpointer,
    }

    log.info(
        "Agent config parts built",
        extra={
            "name": config.agent.name,
            "model": config.model.name,
            "provider": config.model.provider,
            "mode": mode,
            "tools": len(tools),
            "skills": len(skills_paths),
            "sub_agents": len(sub_agents),
        },
    )

    return parts

File: packages/template/src/runtime/agent-config.ts
/**
 * Agent Config Builder
 *
 * Composes the configuration parts passed to deepagents' `createDeepAgent()`:
 * model, system prompt, tools, skills, memory, subagents, permissions,
 * interrupt-on, checkpointer, and the middleware chain. Shared by both the
 * standalone agent factory and the ACP server.
 */

import { type AnyBackendProtocol, type FilesystemPermission, createMemoryMiddleware } from "deepagents";
import type { StructuredTool } from "@langchain/core/tools";
import type { AgentMiddleware } from "langchain";
import { MemorySaver } from "@langchain/langgraph-checkpoint";
import type { BaseCheckpointSaver } from "@langchain/langgraph-checkpoint";
import { type AppConfig, type ACPSessionConfig } from "./config/config-loader.js";
import { createStuckLoopMiddleware } from "./middleware/stuck-loop.js";
import { createPeriodicReminderMiddleware } from "./middleware/periodic-reminder.js";
import { createCostTrackingMiddleware } from "./middleware/cost-tracking.js";
import { createFsPathResolver } from "./middleware/fs-path-resolver.js";
import { createCompactionMiddleware } from "./middleware/compaction.js";
import { createEvictionMiddleware } from "./middleware/eviction.js";
import { createHookMiddleware, getHooks, registerConfiguredHooks } from "../app/hooks/index.js";
import { createProtectedPathsMiddleware } from "./middleware/protected-paths.js";
import { createHarnessLifecycleMiddleware } from "./middleware/harness-lifecycle.js";
import { resolveModel, resolveSummarizerModel } from "./model.js";
import { resolveSystemPrompt, withRuntimeContextPrompt } from "./prompt.js";
import { discoverMemoryFiles, discoverSubAgents, resolveSkillsPaths } from "./discovery.js";
import { resolveSandboxPolicy, buildPermissions, buildInterruptOn, toAbsoluteDenyGlob } from "./permissions.js";

/**
 * Build common agent configuration parts used by both agent-factory and acp-server.
 * Returns an object with all the composed config fields needed for createDeepAgent().
 *
 * @param backend - FilesystemBackend for memory/skills middleware. Required for
 *                  addCacheControl and explicit memory middleware creation.
 */
export function buildAgentConfigParts(
  config: AppConfig,
  sessionConfig: ACPSessionConfig | undefined,
  workspaceRoot: string,
  tools: StructuredTool[],
  backend?: AnyBackendProtocol,
  /**
   * Checkpointing strategy. Pass `false` for one-shot / REPL invocations
   * (no `thread_id` is provided → a checkpointer would throw on
   * `MemorySaver.put`). Pass `true` or a `BaseCheckpointSaver` instance
   * for ACP / long-running session flows that need HITL pause/resume.
   * Defaults to `true` (with a fresh in-memory `MemorySaver`) for
   * backward compatibility — `DeepAgentsServer` in ACP mode overrides
   * this with its own checkpointer.
   */
  checkpointer: true | false | BaseCheckpointSaver = new MemorySaver()
) {
  // Build custom middleware array from config
  const middleware: AgentMiddleware[] = [];
  const mwConfig = config.middleware;
  registerConfiguredHooks(config.hooks, workspaceRoot);

  middleware.push(createHarnessLifecycleMiddleware());

  // Memory middleware — explicitly created with addCacheControl for Anthropic prompt caching.
  // Falls back to the `memory` shortcut parameter when no backend is provided.
  const memoryPaths = discoverMemoryFiles(workspaceRoot, config.agent.includeWorkspaceInstructions);
  if (backend && memoryPaths.length > 0) {
    middleware.push(createMemoryMiddleware({
      backend,
      sources: memoryPaths,
      addCacheControl: config.memory.addCacheControl && config.model.provider === "anthropic",
    }));
  }

  if (mwConfig.stuckLoopDetection.enabled) {
    middleware.push(createStuckLoopMiddleware({
      threshold: mwConfig.stuckLoopDetection.threshold,
      mode: mwConfig.stuckLoopDetection.mode,
    }));
  }

  // Resolve workspace-relative paths ("/test.txt") to absolute paths for ACP clients
  middleware.push(createFsPathResolver(workspaceRoot));

  if (mwConfig.periodicReminder.enabled) {
    middleware.push(createPeriodicReminderMiddleware({
      firstAt: mwConfig.periodicReminder.firstAt,
      every: mwConfig.periodicReminder.every,
    }));
  }

  if (mwConfig.costTracking.enabled) {
    middleware.push(createCostTrackingMiddleware({
      warnAtTokens: mwConfig.costTracking.warnAtTokens,
    }));
  }

  // Context compaction — compress old messages when approaching context limit
  if (config.compaction.enabled) {
    middleware.push(createCompactionMiddleware({
      config: config.compaction,
      modelName: config.model.name,
      summarizer: resolveSummarizerModel(config),
    }));
  }

  // Large output eviction — write oversized tool results to backend
  if (config.eviction.enabled && backend) {
    middleware.push(createEvictionMiddleware({
      config: config.eviction,
      backend: backend as unknown as { write(path: string, content: string): Promise<void> },
    }));
  }

  // Hooks middleware — always included when hooks are registered.
  // Note: hooks registry is module-level (shared across agents in the same process).
  if (getHooks("pre_tool_use").length > 0 || getHooks("post_tool_use").length > 0 ||
      getHooks("before_model").length > 0 || getHooks("after_model").length > 0) {
    middleware.push(createHookMiddleware());
  }

  // Discover subagents from .agents/agents/ directories
  const discoveredSubAgents = discoverSubAgents(config, workspaceRoot);

  // ── Mode-based overrides ─────────────────────────────────
  const mode = config.permissions.mode;
  const sandbox = resolveSandboxPolicy(config);
  let interruptOn: Record<string, boolean>;
  let permissions: FilesystemPermission[];
  let systemPrompt = withRuntimeContextPrompt(
    resolveSystemPrompt(config, sessionConfig, workspaceRoot),
    workspaceRoot
  );

  // Protected paths middleware: wraps the tool call regardless of mode.
  // Needed because `DeepAgentsServer` in ACP mode drops the `permissions`
  // field from `createDeepAgent`, so the `permissions` array alone
  // cannot protect files. Gated only on the sandbox's denied paths.
  if (sandbox.deniedWritePaths.length > 0) {
    const deniedGlobs = sandbox.deniedWritePaths.map((d) => toAbsoluteDenyGlob(d, workspaceRoot));
    middleware.push(createProtectedPathsMiddleware({ deniedGlobs }));
  }

  if (mode === "yolo") {
    // No HITL. Path restrictions are enforced by the protected-paths
    // middleware (pushed above, regardless of mode) — the `permissions`
    // array is kept here for non-ACP callers and is dropped by
    // DeepAgentsServer in ACP mode anyway.
    interruptOn = {};
    permissions = sandbox.profile === "open"
      ? [{ operations: ["read", "write"], paths: ["/**"], mode: "allow" as const }]
      : buildPermissions(config, workspaceRoot);
  } else if (mode === "plan") {
    // HITL on writes + planning preamble
    interruptOn = buildInterruptOn(config.permissions.interruptOn);
    permissions = buildPermissions(config, workspaceRoot);
    const planPreamble = `## Planning Mode
Before making any changes, you MUST:
1. Present a clear plan of what you intend to do
2. Wait for user approval
3. Only then proceed with execution

`;
    systemPrompt = planPreamble + systemPrompt;
  } else {
    // "ask" — default: HITL on writes
    interruptOn = buildInterruptOn(config.permissions.interruptOn);
    permissions = buildPermissions(config, workspaceRoot);
  }

  return {
    model: resolveModel(config),
    systemPrompt,
    tools,
    skills: resolveSkillsPaths(config),
    // When backend is provided, memory is handled by explicit middleware above.
    // Otherwise, fall back to the shortcut parameter (no addCacheControl).
    memory: backend ? undefined : (memoryPaths.length > 0 ? memoryPaths : undefined),
    subagents: discoveredSubAgents.length > 0 ? discoveredSubAgents : undefined,
    permissions,
    interruptOn,
    // REPL/one-shot pass `false` here. ACP (DeepAgentsServer) passes a
    // custom checkpointer that supersedes the value returned from
    // buildAgentConfigParts, so the default `true` keeps the existing
    // ACP behavior.
    checkpointer: checkpointer === true ? new MemorySaver() : checkpointer,
    middleware,
  };
}
